ksy_schema.json

```diff
@@ -216,7 +224,29 @@
             }
           }
         },
-        "license": { "type": "string" },
+        "tags": {
+          "type": "array",
+          "description": "list of some identifiers that can give additional information about format. `lowercase-kebab-case` is preferred for names, as well as keeping the list sorted. You can find more examples in existing schemas at https://formats.kaitai.io/. Should be used only at top-level",
+          "items": { "type": "string" },
+          "examples": [
+            "android",
+            "archive",
+            "database",
+            "dos",
+            "executable",
+            "filesystem",
+            "firmware",
+            "linux",
+            "log",
+            "macos",
+            "media",
+            "windows"
+          ]
+        },
+        "license": {
+          "type": "string",
+          "description": "[SPDX expression](https://spdx.github.io/spdx-spec/appendix-IV-SPDX-license-expressions/)"
```

Comment:
I'm not sure you need to explain the meaning of the license field. What can it be except recursive self-determination?

Reply:
Sorry for not being clear. I'd like the `description` to answer some basic questions about the `license` key in KSY and also to touch on some legal stuff a bit. I'll share here some points that I'd find useful to put there.

https://github.com/kaitai-io/kaitai_struct_formats/pull/39#issuecomment-303241899:

> As for licensing — GPLv2.0 in the footer of id3.org refers to licensing of MoinMoin wiki, an engine that runs their web site. It totally has nothing to do with site contents per se (i.e. with ID3 technology), and _even if it had anything, your work is your work, it is an implementation of that of specification, you're free to license it as you see fit_.

An important point is that the `license` key identifies the license of the KSY implementation, and implementers are free to decide for themselves which license they want, regardless of what license the reference format specification/documentation (in PDF, for example) uses. In other words, one doesn't infringe the copyright of the original spec just by writing a KSY spec according to it and releasing the KSY under own license.

Second, `license` key is required for all KSY specs in KSF.

https://github.com/kaitai-io/kaitai_struct_compiler#libraries-used:

> Note that these clauses only apply only to compiler itself, not `.ksy` input files that one supplies in normal process of compilation, nor to compiler's output files — that consitutes normal usage process_ and you obviously keep copyright to both.

In short, I think it's good to mention that the KSC's output files are licensed under the same license as the input KSY spec. However, I'd avoid the phrase _you obviously keep copyright to both_, because that suggests that the reader is a copyright holder, and that might not be the case.

Maybe @GreyCat could come up with something to add here?

Reply:
> In short, I think it's good to mention that the KSC's output files are licensed under the same license as the input KSY spec.

And this is even true? I mean, I'm not a lawer, but is this property is followed from something? Can output files have their own license?

Reply:
@Mingun:

> > In short, I think it's good to mention that the KSC's output files are licensed under the same license as the input KSY spec.
> 
> And this is even true? I mean, I'm not a lawer, but is this property is followed from something?

Yes, see FAQ about the GPL licenses (https://www.gnu.org/licenses/gpl-faq.en.html#GPLOutput):

> **Is there some way that I can GPL the output people get from use of my program? For example, if my program is used to develop hardware designs, can I require that these designs must be free?** ([#GPLOutput](https://www.gnu.org/licenses/gpl-faq.en.html#GPLOutput))
>
> In general this is legally impossible; copyright law does not give you any say in the use of the output people make from their data using your program. If the user uses your program to enter or convert her own data, the copyright on the output belongs to her, not you. More generally, when a program translates its input into some other form, the copyright status of the output inherits that of the input it was generated from.

---

> Can output files have their own license?

Do you mean that we would release KSC under the terms that its output files would subject to some license that we want, instead of borrowing the license that the input file subjects to? No, that's not possible. Output files can't have _their own license_, see the above excerpt from the GPL FAQ.

It makes a lot of sense to me. Imagine you have the following YAML file released under a license XYZ:

```yaml
meta:
  id: hello_world
  license: XYZ
seq:
  - id: one
    type: u1
```

Well, you can see that it's actually a KSY spec, but let's treat it as a general YAML file. Then there's a tool for converting YAML to JSON (e.g. https://yaml-online-parser.appspot.com/). So we use it and we get the following output file:

```json
{
  "meta": {
    "id": "hello_world", 
    "license": "XYZ"
  }, 
  "seq": [
    {
      "type": "u1", 
      "id": "one"
    }
  ]
}
```

The question is, what license does this JSON output subject to? Well, the only correct answer is the license XYZ, because it's just a derived work from the input YAML. The same argument applies to a KSY spec and the KS compiler. The output parser is just a "trivial" transcription of the original .ksy spec.

Reply:
Ok, got it! Actually, I'm interested in situation, when you write your program under license _X_ and use the generated code from KSY, and KSY file under license _Y_.